
# Acquiring data for the last 10 years using longitude & latitude values for shanghai
## Temperature,	Wind Direction,	Wind Speed,	Humidity & Pressure

In [ ]:
!pip install meteostat

In [ ]:
from meteostat import Point, Hourly
import pandas as pd
from datetime import datetime, timedelta

# Define the location for Shanghai
shanghai = Point(31.2304, 121.4737)

# Specify the date range for the past 30 days
end_date = datetime.now()
start_date = pd.to_datetime('2014-01-01')

# Fetch hourly historical data
data = Hourly(shanghai, start=start_date, end=end_date)

# Convert data to a DataFrame
df = data.fetch()

# Display the DataFrame
print(df.head())

# Optionally, save to CSV
df.to_csv('shanghai_weather_data.csv')

In [40]:
data=pd.read_csv('shanghai_weather_data.csv')
data.head()

time  temp  dwpt  rhum  prcp  snow   wdir  wspd  wpgt  pres  \
0  2014-01-01 00:00:00   4.0  -4.0  56.0   NaN   NaN  230.0   7.2   NaN   NaN   
1  2014-01-01 01:00:00   7.0  -4.1  45.0   NaN   NaN  230.0  10.8   NaN   NaN   
2  2014-01-01 02:00:00  10.0  -5.2  34.0   NaN   NaN  240.0  14.4   NaN   NaN   
3  2014-01-01 03:00:00  13.0  -5.1  28.0   NaN   NaN  260.0  14.4   NaN   NaN   
4  2014-01-01 04:00:00  14.0  -6.8  23.0   NaN   NaN  250.0  18.0   NaN   NaN   

   tsun  coco  
0   NaN   NaN  
1   NaN   NaN  
2   NaN   NaN  
3   NaN   NaN  
4   NaN   NaN

## filtering needed columns and renaming them for easy understanding

In [5]:
data_filtered = data[['time', 'temp', 'wdir', 'wspd', 'rhum', 'pres']].rename(columns={
    'time': 'Timestamp',
    'temp': 'Temperature',
    'wdir': 'Wind Direction',
    'wspd': 'Wind Speed',
    'rhum': 'Humidity',
    'pres': 'Pressure'
})


In [6]:
data_filtered.isnull().sum()

Timestamp             0
Temperature          62
Wind Direction     2766
Wind Speed          171
Humidity             62
Pressure          14960
dtype: int64

In [7]:
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96064 entries, 0 to 96063
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Timestamp       96064 non-null  object 
 1   Temperature     96002 non-null  float64
 2   Wind Direction  93298 non-null  float64
 3   Wind Speed      95893 non-null  float64
 4   Humidity        96002 non-null  float64
 5   Pressure        81104 non-null  float64
dtypes: float64(5), object(1)
memory usage: 4.4+ MB


In [8]:
import pandas as pd

# Assuming you already have your DataFrame as 'df_hourly'
# Calculate percentage of missing values
missing_percentages = data_filtered.isnull().mean() * 100

# Filter for columns with missing data
missing_data_columns = missing_percentages[missing_percentages > 0]

# Display the results
print("Percentage of missing values in columns with missing data:")
print(missing_data_columns)

Percentage of missing values in columns with missing data:
Temperature        0.064540
Wind Direction     2.879330
Wind Speed         0.178006
Humidity           0.064540
Pressure          15.572951
dtype: float64


# Checking rows with missing values to find a way for imputation

In [9]:
null_wind_speed=data_filtered[data_filtered['Wind Speed'].isnull()]
null_wind_speed.head(50)


Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
190    2014-01-08 22:00:00          NaN             NaN         NaN       NaN   
290    2014-01-13 02:00:00          NaN             NaN         NaN       NaN   
329    2014-01-14 17:00:00          NaN             NaN         NaN       NaN   
1062   2014-02-14 06:00:00          NaN             NaN         NaN       NaN   
1091   2014-02-15 11:00:00          NaN             NaN         NaN       NaN   
1092   2014-02-15 12:00:00          NaN             NaN         NaN       NaN   
1166   2014-02-18 14:00:00          NaN             NaN         NaN       NaN   
3014   2014-05-06 14:00:00          NaN             NaN         NaN       NaN   
3015   2014-05-06 15:00:00          NaN             NaN         NaN       NaN   
3038   2014-05-07 14:00:00          NaN             NaN         NaN       NaN   
3139   2014-05-11 19:00:00          NaN             NaN         NaN       NaN   
3140   2014-05-11 20:00:00          NaN             NaN         NaN       NaN   
3141   2014-05-11 21:00:00          NaN             NaN         NaN       NaN   
3307   2014-05-18 19:00:00          NaN             NaN         NaN       NaN   
3903   2014-06-12 15:00:00          NaN             NaN         NaN       NaN   
3908   2014-06-12 20:00:00          NaN             NaN         NaN       NaN   
3933   2014-06-13 21:00:00          NaN             NaN         NaN       NaN   
4501   2014-07-07 13:00:00          NaN             NaN         NaN       NaN   
4502   2014-07-07 14:00:00          NaN             NaN         NaN       NaN   
4503   2014-07-07 15:00:00          NaN             NaN         NaN       NaN   
4621   2014-07-12 13:00:00          NaN             NaN         NaN       NaN   
4622   2014-07-12 14:00:00          NaN             NaN         NaN       NaN   
4623   2014-07-12 15:00:00          NaN             NaN         NaN       NaN   
4624   2014-07-12 16:00:00          NaN             NaN         NaN       NaN   
4625   2014-07-12 17:00:00          NaN             NaN         NaN       NaN   
4626   2014-07-12 18:00:00          NaN             NaN         NaN       NaN   
4627   2014-07-12 19:00:00          NaN             NaN         NaN       NaN   
4628   2014-07-12 20:00:00          NaN             NaN         NaN       NaN   
4629   2014-07-12 21:00:00          NaN             NaN         NaN       NaN   
4630   2014-07-12 22:00:00          NaN             NaN         NaN       NaN   
4632   2014-07-13 00:00:00          NaN             NaN         NaN       NaN   
4635   2014-07-13 03:00:00          NaN             NaN         NaN       NaN   
4641   2014-07-13 09:00:00          NaN             NaN         NaN       NaN   
5149   2014-08-03 13:00:00          NaN             NaN         NaN       NaN   
5902   2014-09-03 22:00:00          NaN             NaN         NaN       NaN   
6256   2014-09-18 16:00:00          NaN             NaN         NaN       NaN   
6351   2014-09-22 15:00:00          NaN             NaN         NaN       NaN   
6543   2014-09-30 15:00:00          NaN             NaN         NaN       NaN   
6832   2014-10-12 16:00:00          NaN             NaN         NaN       NaN   
6833   2014-10-12 17:00:00          NaN             NaN         NaN       NaN   
6834   2014-10-12 18:00:00          NaN             NaN         NaN       NaN   
6835   2014-10-12 19:00:00          NaN             NaN         NaN       NaN   
7864   2014-11-24 16:00:00          NaN             NaN         NaN       NaN   
7911   2014-11-26 15:00:00          NaN             NaN         NaN       NaN   
8252   2014-12-10 20:00:00          NaN             NaN         NaN       NaN   
8255   2014-12-10 23:00:00          NaN             NaN         NaN       NaN   
8331   2014-12-14 03:00:00          NaN             NaN         NaN       NaN   
10950  2015-04-02 06:00:00          NaN             NaN         NaN       NaN   
11094  2015-04-08 06:00:00          NaN         

In [10]:
null_temp=data_filtered[(data_filtered['Temperature'].isnull())]
null_temp.head(50)

Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
190    2014-01-08 22:00:00          NaN             NaN         NaN       NaN   
290    2014-01-13 02:00:00          NaN             NaN         NaN       NaN   
329    2014-01-14 17:00:00          NaN             NaN         NaN       NaN   
1062   2014-02-14 06:00:00          NaN             NaN         NaN       NaN   
1091   2014-02-15 11:00:00          NaN             NaN         NaN       NaN   
1092   2014-02-15 12:00:00          NaN             NaN         NaN       NaN   
1166   2014-02-18 14:00:00          NaN             NaN         NaN       NaN   
3014   2014-05-06 14:00:00          NaN             NaN         NaN       NaN   
3015   2014-05-06 15:00:00          NaN             NaN         NaN       NaN   
3038   2014-05-07 14:00:00          NaN             NaN         NaN       NaN   
3139   2014-05-11 19:00:00          NaN             NaN         NaN       NaN   
3140   2014-05-11 20:00:00          NaN             NaN         NaN       NaN   
3141   2014-05-11 21:00:00          NaN             NaN         NaN       NaN   
3307   2014-05-18 19:00:00          NaN             NaN         NaN       NaN   
3903   2014-06-12 15:00:00          NaN             NaN         NaN       NaN   
3908   2014-06-12 20:00:00          NaN             NaN         NaN       NaN   
3933   2014-06-13 21:00:00          NaN             NaN         NaN       NaN   
4501   2014-07-07 13:00:00          NaN             NaN         NaN       NaN   
4502   2014-07-07 14:00:00          NaN             NaN         NaN       NaN   
4503   2014-07-07 15:00:00          NaN             NaN         NaN       NaN   
4621   2014-07-12 13:00:00          NaN             NaN         NaN       NaN   
4622   2014-07-12 14:00:00          NaN             NaN         NaN       NaN   
4623   2014-07-12 15:00:00          NaN             NaN         NaN       NaN   
4624   2014-07-12 16:00:00          NaN             NaN         NaN       NaN   
4625   2014-07-12 17:00:00          NaN             NaN         NaN       NaN   
4626   2014-07-12 18:00:00          NaN             NaN         NaN       NaN   
4627   2014-07-12 19:00:00          NaN             NaN         NaN       NaN   
4628   2014-07-12 20:00:00          NaN             NaN         NaN       NaN   
4629   2014-07-12 21:00:00          NaN             NaN         NaN       NaN   
4630   2014-07-12 22:00:00          NaN             NaN         NaN       NaN   
4632   2014-07-13 00:00:00          NaN             NaN         NaN       NaN   
4635   2014-07-13 03:00:00          NaN             NaN         NaN       NaN   
4641   2014-07-13 09:00:00          NaN             NaN         NaN       NaN   
5149   2014-08-03 13:00:00          NaN             NaN         NaN       NaN   
5902   2014-09-03 22:00:00          NaN             NaN         NaN       NaN   
6256   2014-09-18 16:00:00          NaN             NaN         NaN       NaN   
6351   2014-09-22 15:00:00          NaN             NaN         NaN       NaN   
6543   2014-09-30 15:00:00          NaN             NaN         NaN       NaN   
6832   2014-10-12 16:00:00          NaN             NaN         NaN       NaN   
6833   2014-10-12 17:00:00          NaN             NaN         NaN       NaN   
6834   2014-10-12 18:00:00          NaN             NaN         NaN       NaN   
6835   2014-10-12 19:00:00          NaN             NaN         NaN       NaN   
7864   2014-11-24 16:00:00          NaN             NaN         NaN       NaN   
7911   2014-11-26 15:00:00          NaN             NaN         NaN       NaN   
8252   2014-12-10 20:00:00          NaN             NaN         NaN       NaN   
8255   2014-12-10 23:00:00          NaN             NaN         NaN       NaN   
8331   2014-12-14 03:00:00          NaN             NaN         NaN       NaN   
8889   2015-01-06 09:00:00          NaN            20.0        21.6       NaN   
10137  2015-02-27 09:00:00          NaN         

In [11]:
print(data_filtered.iloc[180:200,:])

               Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
180  2014-01-08 12:00:00          5.0           360.0        25.2      75.0   
181  2014-01-08 13:00:00          4.0           360.0        25.2      75.0   
182  2014-01-08 14:00:00          3.0           360.0        14.4      75.0   
183  2014-01-08 15:00:00          3.0           350.0        14.4      75.0   
184  2014-01-08 16:00:00          3.0           350.0        14.4      75.0   
185  2014-01-08 17:00:00          3.0           350.0        18.0      70.0   
186  2014-01-08 18:00:00          2.0           360.0        21.6      75.0   
187  2014-01-08 19:00:00          2.0           350.0        14.4      69.0   
188  2014-01-08 20:00:00          2.0           360.0        14.4      75.0   
189  2014-01-08 21:00:00          1.0           350.0        10.8      75.0   
190  2014-01-08 22:00:00          NaN             NaN         NaN       NaN   
191  2014-01-08 23:00:00          1.0           330.

# Above EDA shows missing values with small and bigger gaps in data with respect to time.
## We impute gaps with upto 2 missing value using interpolation
### i)for gradual factors like temperature we use linear interpolation 
### ii)quadratic for Humidity to capture some curve in the data pattern.
### iii)cubic for Wind_Speed to better capture potential fluctuations.
### iv)nearest for Pressure since this variable may not change drastically in the short term.




In [ ]:
!pip install scipy

In [14]:
# Step 1: Interpolate small gaps (up to 2 consecutive missing values) with specified methods
data_filtered['Temperature'] = data_filtered['Temperature'].interpolate(method='linear', limit=2)
data_filtered['Humidity'] = data_filtered['Humidity'].interpolate(method='quadratic', limit=2)
data_filtered['Wind Speed'] = data_filtered['Wind Speed'].interpolate(method='cubic', limit=2)
data_filtered['Pressure'] = data_filtered['Pressure'].interpolate(method='nearest', limit=2)

## Using a 7-day rolling average to fill larger gaps, which helps capture broader patterns and minimizes the effect of missing data.

In [15]:
data_filtered['Temperature'] = data_filtered['Temperature'].fillna(data_filtered['Temperature'].rolling(window=24*7, min_periods=1).mean())
data_filtered['Humidity'] = data_filtered['Humidity'].fillna(data_filtered['Humidity'].rolling(window=24*7, min_periods=1).mean())
data_filtered['Wind Speed'] = data_filtered['Wind Speed'].fillna(data_filtered['Wind Speed'].rolling(window=24*7, min_periods=1).mean())
data_filtered['Pressure'] = data_filtered['Pressure'].fillna(data_filtered['Pressure'].rolling(window=24*7, min_periods=1).mean())

## Checking for null values after imputation

In [16]:
data_filtered.isnull().sum()

Timestamp             0
Temperature           0
Wind Direction     2766
Wind Speed            0
Humidity              0
Pressure          14010
dtype: int64

## presure still has missing values

In [17]:
pressure_null=data_filtered[(data_filtered['Pressure'].isnull())]
pressure_null.head(50)

Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
0   2014-01-01 00:00:00          4.0           230.0         7.2      56.0   
1   2014-01-01 01:00:00          7.0           230.0        10.8      45.0   
2   2014-01-01 02:00:00         10.0           240.0        14.4      34.0   
3   2014-01-01 03:00:00         13.0           260.0        14.4      28.0   
4   2014-01-01 04:00:00         14.0           250.0        18.0      23.0   
5   2014-01-01 05:00:00         16.0           260.0        18.0      17.0   
6   2014-01-01 06:00:00         16.0           250.0        18.0      18.0   
7   2014-01-01 07:00:00         16.0           270.0        18.0      15.0   
8   2014-01-01 08:00:00         15.0           280.0        14.4      18.0   
9   2014-01-01 09:00:00         13.0           250.0         7.2      22.0   
10  2014-01-01 10:00:00         12.0             NaN         3.6      24.0   
11  2014-01-01 11:00:00         10.0             NaN         0.0      30.0   
12  2014-01-01 12:00:00         10.0           170.0         7.2      32.0   
13  2014-01-01 13:00:00          9.0             NaN         3.6      37.0   
14  2014-01-01 14:00:00          8.0             NaN         3.6      39.0   
15  2014-01-01 15:00:00          7.0             NaN         0.0      42.0   
16  2014-01-01 16:00:00          5.0           170.0         7.2      52.0   
17  2014-01-01 17:00:00          5.0             NaN         0.0      56.0   
18  2014-01-01 18:00:00          3.0             NaN         0.0      60.0   
19  2014-01-01 19:00:00          3.0             NaN         0.0      65.0   
20  2014-01-01 20:00:00          2.0             NaN         0.0      75.0   
21  2014-01-01 21:00:00          2.0             NaN         0.0      69.0   
22  2014-01-01 22:00:00          2.0             NaN         0.0      69.0   
23  2014-01-01 23:00:00          2.0             NaN         0.0      69.0   
24  2014-01-02 00:00:00          6.0            70.0         3.6      65.0   
25  2014-01-02 01:00:00         11.0           130.0        10.8      47.0   
26  2014-01-02 02:00:00         13.0           130.0        14.4      33.0   
27  2014-01-02 03:00:00         15.0           110.0        14.4      29.0   
28  2014-01-02 04:00:00         16.0           120.0        14.4      23.0   
29  2014-01-02 05:00:00         17.0           120.0        14.4      16.0   
30  2014-01-02 06:00:00         18.0           150.0        18.0      18.0   
31  2014-01-02 07:00:00         17.0           150.0        21.6      22.0   
32  2014-01-02 08:00:00         16.0           150.0        18.0      31.0   
33  2014-01-02 09:00:00         15.0           130.0        18.0      39.0   
34  2014-01-02 10:00:00         13.0           120.0        18.0      51.0   
35  2014-01-02 11:00:00         12.0           120.0        14.4      58.0   
36  2014-01-02 12:00:00         12.0           120.0        14.4      62.0   
37  2014-01-02 13:00:00         11.0           130.0        14.4      67.0   
38  2014-01-02 14:00:00         10.0           130.0        10.8      71.0   
39  2014-01-02 15:00:00         11.0           140.0        14.4      71.0   
40  2014-01-02 16:00:00         11.0           150.0        14.4      71.0   
41  2014-01-02 17:00:00         11.0             NaN         3.6      71.0   
42  2014-01-02 18:00:00         10.0           160.0         3.6      76.0   
43  2014-01-02 19:00:00          9.0             NaN         0.0      81.0   
44  2014-01-02 20:00:00          9.0             NaN         0.0      81.0   
45  2014-01-02 21:00:00          9.0             NaN         0.0      81.0   
46  2014-01-02 22:00:00          9.0           340.0         3.6      81.0   
47  2014-01-02 23:00:00          9.0           280.0         3.6      81.0   
48  2014-01-03 00:00:00         10.0             NaN         3.6      76.0   
49  2014-01-03 01:00:00         11.0           300.0        14.4      71.0   

    Pressure  
0        NaN  
1    

In [18]:
variance = data_filtered['Pressure'].var()
std_dev = data_filtered['Pressure'].std()
print("variance:",variance, "standard deviation:", std_dev)

variance: 95.3852805969902 standard deviation: 9.766538823810112


In [19]:
print(data_filtered.iloc[49:100,:])

              Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
49  2014-01-03 01:00:00         11.0           300.0        14.4      71.0   
50  2014-01-03 02:00:00         12.0           320.0        18.0      62.0   
51  2014-01-03 03:00:00         13.0           320.0        18.0      54.0   
52  2014-01-03 04:00:00         15.0           320.0        21.6      41.0   
53  2014-01-03 05:00:00         16.0           330.0        18.0      27.0   
54  2014-01-03 06:00:00         17.0           340.0        21.6      22.0   
55  2014-01-03 07:00:00         16.0           340.0        18.0      18.0   
56  2014-01-03 08:00:00         15.0           330.0        18.0      20.0   
57  2014-01-03 09:00:00         14.0           320.0        14.4      23.0   
58  2014-01-03 10:00:00         12.0           340.0        10.8      28.0   
59  2014-01-03 11:00:00         12.0           320.0        10.8      30.0   
60  2014-01-03 12:00:00         12.0           280.0        10.8

In [41]:
pressure_notnull=data_filtered[(data_filtered['Pressure'].notnull())]
pressure_notnull.head()             

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2014-01-01 00:00:00          4.0           230.0         7.2      56.0   
2014-01-01 01:00:00          7.0           230.0        10.8      45.0   
2014-01-01 02:00:00         10.0           240.0        14.4      34.0   
2014-01-01 03:00:00         13.0           260.0        14.4      28.0   
2014-01-01 04:00:00         14.0           250.0        18.0      23.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2014-01-01 00:00:00    1026.0 -0.766044 -0.642788                        230.0  
2014-01-01 01:00:00    1026.0 -0.766044 -0.642788                        230.0  
2014-01-01 02:00:00    1026.0 -0.866025 -0.500000                        240.0  
2014-01-01 03:00:00    1026.0 -0.984808 -0.173648                        260.0  
2014-01-01 04:00:00    1026.0 -0.939693 -0.342020                        250.0

## pressure doesnot vary frequently and has large gaps so we impute the missing value with median/mean of that specific month 

In [21]:
data_filtered['Timestamp'] = pd.to_datetime(data_filtered['Timestamp'])

# Fill missing values in 'Pressure' based on monthly median
data_filtered['Pressure'] = data_filtered['Pressure'].fillna(
    data_filtered['Pressure'].groupby(data_filtered['Timestamp'].dt.month).transform('median')
)


## only wind direction has null values now

In [22]:
data_filtered.isnull().sum()

Timestamp            0
Temperature          0
Wind Direction    2766
Wind Speed           0
Humidity             0
Pressure             0
dtype: int64

In [42]:
data_filtered.tail(10)

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2024-12-16 06:00:00          9.1           262.0         7.4      31.0   
2024-12-16 07:00:00          8.9           264.0         7.4      33.0   
2024-12-16 08:00:00          8.8           267.0         5.5      35.0   
2024-12-16 09:00:00          8.7           271.0         5.5      37.0   
2024-12-16 10:00:00          8.1           251.0         5.5      40.0   
2024-12-16 11:00:00          7.6           229.0         3.7      43.0   
2024-12-16 12:00:00          7.0           210.0         3.7      47.0   
2024-12-16 13:00:00          6.6           211.0         3.7      50.0   
2024-12-16 14:00:00          6.2           213.0         3.7      54.0   
2024-12-16 15:00:00          5.8           215.0         5.5      59.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2024-12-16 06:00:00    1024.8 -0.990268 -0.139173                        262.0  
2024-12-16 07:00:00    1024.8 -0.994522 -0.104528                        264.0  
2024-12-16 08:00:00    1024.9 -0.998630 -0.052336                        267.0  
2024-12-16 09:00:00    1024.9 -0.999848  0.017452                        271.0  
2024-12-16 10:00:00    1025.2 -0.945519 -0.325568                        251.0  
2024-12-16 11:00:00    1025.5 -0.754710 -0.656059                        229.0  
2024-12-16 12:00:00    1025.7 -0.500000 -0.866025                        210.0  
2024-12-16 13:00:00    1025.9 -0.515038 -0.857167                        211.0  
2024-12-16 14:00:00    1026.0 -0.544639 -0.838671                        213.0  
2024-12-16 15:00:00    1026.1 -0.573576 -0.819152                        215.0

### Filtering rows with 0 wind speed and corresponding wind direction

In [24]:
# 0 wind speed but wind direction values available
zero_speed_direction= data_filtered[(data_filtered['Wind Speed'] == 0) & (data_filtered['Wind Direction'].notnull())]
#there are 559 rows with no wind speed but direction mentioned
zero_speed_direction[['Wind Speed', 'Wind Direction']].describe()


Wind Speed  Wind Direction
count       572.0      572.000000
mean          0.0       53.590909
std           0.0      102.406236
min           0.0        0.000000
25%           0.0        0.000000
50%           0.0        0.000000
75%           0.0       61.250000
max           0.0      358.000000

## Filtering rows with null wind direction
### many instances with wind speed equal to 0 and less wind speeds 

In [43]:
null_direction=data_filtered[data_filtered['Wind Direction'].isnull()]
null_direction.head(10)

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2014-01-01 15:00:00          7.0             NaN         0.0      42.0   
2014-01-01 19:00:00          3.0             NaN         0.0      65.0   
2014-01-01 20:00:00          2.0             NaN         0.0      75.0   
2014-01-01 21:00:00          2.0             NaN         0.0      69.0   
2014-01-01 22:00:00          2.0             NaN         0.0      69.0   
2014-01-01 23:00:00          2.0             NaN         0.0      69.0   
2014-01-02 21:00:00          9.0             NaN         0.0      81.0   
2014-01-05 05:00:00         10.0             NaN         3.6      54.0   
2014-01-05 06:00:00         11.0             NaN         3.6      47.0   
2014-01-05 07:00:00         11.0             NaN         3.6      47.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2014-01-01 15:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 19:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 20:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 21:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 22:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 23:00:00    1026.0       NaN       NaN                          NaN  
2014-01-02 21:00:00    1026.0       NaN       NaN                          NaN  
2014-01-05 05:00:00    1026.0       NaN       NaN                          NaN  
2014-01-05 06:00:00    1026.0       NaN       NaN                          NaN  
2014-01-05 07:00:00    1026.0       NaN       NaN                          NaN

## examining rows with wind direction values present to help with imputing missing values

In [44]:
wind_direction_not_null= data_filtered[data_filtered['Wind Direction'].notnull()]
wind_direction_not_null.tail(20)

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2024-12-15 20:00:00          1.0           230.0         7.0      69.0   
2024-12-15 21:00:00          1.0           250.0         4.0      69.0   
2024-12-15 22:00:00          0.0           210.0         7.0      75.0   
2024-12-15 23:00:00          2.0           250.0         4.0      70.0   
2024-12-16 00:00:00          2.2           255.0         3.7      62.0   
2024-12-16 01:00:00          7.0           260.0         7.0      46.0   
2024-12-16 02:00:00          9.0           280.0        11.0      37.0   
2024-12-16 03:00:00         10.0           250.0        14.0      37.0   
2024-12-16 04:00:00          8.5           248.0         7.4      34.0   
2024-12-16 05:00:00          8.8           256.0         7.4      33.0   
2024-12-16 06:00:00          9.1           262.0         7.4      31.0   
2024-12-16 07:00:00          8.9           264.0         7.4      33.0   
2024-12-16 08:00:00          8.8           267.0         5.5      35.0   
2024-12-16 09:00:00          8.7           271.0         5.5      37.0   
2024-12-16 10:00:00          8.1           251.0         5.5      40.0   
2024-12-16 11:00:00          7.6           229.0         3.7      43.0   
2024-12-16 12:00:00          7.0           210.0         3.7      47.0   
2024-12-16 13:00:00          6.6           211.0         3.7      50.0   
2024-12-16 14:00:00          6.2           213.0         3.7      54.0   
2024-12-16 15:00:00          5.8           215.0         5.5      59.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2024-12-15 20:00:00    1027.0 -0.766044 -0.642788                        230.0  
2024-12-15 21:00:00    1027.0 -0.939693 -0.342020                        250.0  
2024-12-15 22:00:00    1028.0 -0.500000 -0.866025                        210.0  
2024-12-15 23:00:00    1028.0 -0.939693 -0.342020                        250.0  
2024-12-16 00:00:00    1028.3 -0.965926 -0.258819                        255.0  
2024-12-16 01:00:00    1028.0 -0.984808 -0.173648                        260.0  
2024-12-16 02:00:00    1028.0 -0.984808  0.173648                        280.0  
2024-12-16 03:00:00    1027.0 -0.939693 -0.342020                        250.0  
2024-12-16 04:00:00    1026.6 -0.927184 -0.374607                        248.0  
2024-12-16 05:00:00    1025.7 -0.970296 -0.241922                        256.0  
2024-12-16 06:00:00    1024.8 -0.990268 -0.139173                        262.0  
2024-12-16 07:00:00    1024.8 -0.994522 -0.104528                        264.0  
2024-12-16 08:00:00    1024.9 -0.998630 -0.052336                        267.0  
2024-12-16 09:00:00    1024.9 -0.999848  0.017452                        271.0  
2024-12-16 10:00:00    1025.2 -0.945519 -0.325568                        251.0  
2024-12-16 11:00:00    1025.5 -0.754710 -0.656059                        229.0  
2024-12-16 12:00:00    1025.7 -0.500000 -0.866025                        210.0  
2024-12-16 13:00:00    1025.9 -0.515038 -0.857167                        211.0  
2024-12-16 14:00:00    1026.0 -0.544639 -0.838671                        213.0  
2024-12-16 15:00:00    1026.1 -0.573576 -0.819152                        215.0

## wind direction is in degrees so directly interpolating them will not take into account the circular data
### the direction is broken into sine and cosine components 

In [27]:
import numpy as np
data_filtered['wdir_sin'] = np.sin(np.radians(data_filtered['Wind Direction']))
data_filtered['wdir_cos'] = np.cos(np.radians(data_filtered['Wind Direction']))

## linearly interpolate values with small gaps i.e 2

In [28]:
data_filtered['wdir_sin'] = data_filtered['wdir_sin'].interpolate(method='linear', limit=2)
data_filtered['wdir_cos'] = data_filtered['wdir_cos'].interpolate(method='linear', limit=2)

In [29]:
# Recalculate wind direction from interpolated sine and cosine
data_filtered['Wind_Direction_Interpolated'] = np.degrees(np.arctan2(data_filtered['wdir_sin'], data_filtered['wdir_cos']))

# Ensure wind direction is between 0 and 360 degrees
data_filtered['Wind_Direction_Interpolated'] = data_filtered['Wind_Direction_Interpolated'] % 360
# Replace original Wind_Direction with interpolated values where Wind_Direction was NaN
data_filtered['Wind Direction'] = data_filtered['Wind Direction'].fillna(data_filtered['Wind_Direction_Interpolated'])


## only 483 null values remain

In [30]:
data_filtered.isnull().sum()

Timestamp                        0
Temperature                      0
Wind Direction                 483
Wind Speed                       0
Humidity                         0
Pressure                         0
wdir_sin                       483
wdir_cos                       483
Wind_Direction_Interpolated    483
dtype: int64

In [31]:
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96064 entries, 0 to 96063
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Timestamp                    96064 non-null  datetime64[ns]
 1   Temperature                  96064 non-null  float64       
 2   Wind Direction               95581 non-null  float64       
 3   Wind Speed                   96064 non-null  float64       
 4   Humidity                     96064 non-null  float64       
 5   Pressure                     96064 non-null  float64       
 6   wdir_sin                     95581 non-null  float64       
 7   wdir_cos                     95581 non-null  float64       
 8   Wind_Direction_Interpolated  95581 non-null  float64       
dtypes: datetime64[ns](1), float64(8)
memory usage: 6.6 MB


## Out of 95053 rows 483 null values remaining with 182 instances having wind speed equal to 0

In [32]:
wind_zero_null_direction=data_filtered[(data_filtered['Wind Direction'].isnull()) & (data_filtered['Wind Speed']==0 )]
wind_zero_null_direction.describe()

Timestamp  Temperature  Wind Direction  Wind Speed  \
count                            182   182.000000             0.0       182.0   
mean   2016-04-10 17:30:59.340659200    10.796703             NaN         0.0   
min              2014-01-01 15:00:00    -4.000000             NaN         0.0   
25%              2015-06-25 03:30:00     4.000000             NaN         0.0   
50%              2016-01-08 19:30:00    10.000000             NaN         0.0   
75%              2017-09-28 02:00:00    16.000000             NaN         0.0   
max              2018-11-28 21:00:00    32.000000             NaN         0.0   
std                              NaN     7.641656             NaN         0.0   

         Humidity     Pressure  wdir_sin  wdir_cos  \
count  182.000000   182.000000       0.0       0.0   
mean    81.802198  1021.523316       NaN       NaN   
min     42.000000  1005.000000       NaN       NaN   
25%     75.250000  1019.000000       NaN       NaN   
50%     82.000000  1022.000000       NaN       NaN   
75%     88.000000  1026.000000       NaN       NaN   
max    100.000000  1036.000000       NaN       NaN   
std     10.877235     6.705822       NaN       NaN   

       Wind_Direction_Interpolated  
count                          0.0  
mean                           NaN  
min                            NaN  
25%                            NaN  
50%                            NaN  
75%                            NaN  
max                            NaN  
std                            NaN

## Aggregating hourly data to monthly data 
### remaining null values in wind direction indicate low wind speed or no wind so we keep them and impute mean of wind direction for that specific month 
### using mean for all other parameters.

In [34]:
data_filtered.set_index('Timestamp', inplace=True)

monthly_data = data_filtered.resample('M').agg({
    'Wind Direction': lambda x: np.degrees(np.arctan2(np.nanmean(np.sin(np.radians(x))), np.nanmean(np.cos(np.radians(x))))) % 360,
    'Wind Speed': 'mean',  # Use mean or other method as preferred
    'Temperature': 'mean',
    'Humidity': 'mean',
    'Pressure': 'mean'
}).where(data_filtered.resample('M').count() >= 20)  # Adjust min_periods as needed


## no null values

In [35]:
monthly_data.isnull().sum()

Wind Direction    0
Wind Speed        0
Temperature       0
Humidity          0
Pressure          0
dtype: int64

In [45]:
monthly_data.tail(15)

Wind Direction  Wind Speed  Temperature   Humidity     Pressure
Timestamp                                                                  
2023-10-31       48.550879   11.081720    21.066801  67.419355  1019.397715
2023-11-30      321.215866   13.935556    14.887500  60.848611  1023.186250
2023-12-31      329.751841   12.502151     7.537634  65.294355  1026.823253
2024-01-31      356.159741   12.846774     6.594086  66.689516  1026.935215
2024-02-29      347.379127   15.516667     6.846264  75.396552  1024.964511
2024-03-31       96.409317   13.397849    12.271909  67.778226  1018.720565
2024-04-30       96.652195   13.155000    17.545833  73.590278  1012.629028
2024-05-31      124.164003   14.695968    21.677419  64.368280  1013.188710
2024-06-30      164.679932   12.019444    24.929167  77.311111  1007.366111
2024-07-31      178.374792   17.010753    31.580645  69.548387  1004.698522
2024-08-31      135.622965   12.851344    31.934140  67.279570  1005.059812
2024-09-30       84.487889   15.679167    28.128889  76.018056  1009.637222
2024-10-31       18.242554   14.990054    20.412634  74.865591  1019.106183
2024-11-30       14.797232   14.293194    15.417639  67.195833  1023.228750
2024-12-31       11.937579   10.410638     9.045213  60.630319  1025.261436

## AIR quality data extracted from a website with monthly data

In [ ]:
!pip install openpyxl

In [48]:
air_qa=pd.read_excel(r'C:\Users\hp\Downloads\timeserie\air_quality.xlsx')

In [49]:
air_qa.head()

month  AQI   scope       Quality Grade  PM2.5  PM10  SO2  NO2     CO   O3
0  2013-12  158  24~468  Moderate pollution    125   140   48   71  1.373   49
1  2014-01  103  30~238   Slightly polluted     77    90   28   54  0.997   58
2  2014-02   74  27~240                good     52    59   17   36  0.796   73
3  2014-03   82  39~202                good     57    84   19   51  0.787  100
4  2014-04   77  50~187                good     53    67   15   47  0.727  108

In [51]:
air_qa.head(20)

month  AQI   scope       Quality Grade  PM2.5  PM10  SO2  NO2     CO  \
0   2013-12  158  24~468  Moderate pollution    125   140   48   71  1.373   
1   2014-01  103  30~238   Slightly polluted     77    90   28   54  0.997   
2   2014-02   74  27~240                good     52    59   17   36  0.796   
3   2014-03   82  39~202                good     57    84   19   51  0.787   
4   2014-04   77  50~187                good     53    67   15   47  0.727   
5   2014-05   96  53~193                good     63    99   17   42  0.777   
6   2014-06   72  37~182                good     44    58   12   35  0.720   
7   2014-07   74  32~167                good     45    63   12   31  0.765   
8   2014-08   67  30~150                good     38    52   13   34  0.768   
9   2014-09   69  32~138                good     35    52   13   32  0.740   
10  2014-10   77  52~137                good     40    68   16   42  0.781   
11  2014-11   74  38~182                good     52    71   18   52  0.843   
12  2014-12   99  39~190                good     72   113   36   68  1.090   
13  2015-01  111  28~256   Slightly polluted     83   108   29   63  1.090   
14  2015-02   90  34~222                good     64    85   20   46  0.946   
15  2015-03   81  37~143                good     54    66   15   48  0.845   
16  2015-04   88  44~166                good     56    81   16   53  0.790   
17  2015-05   77  51~117                good     44    63   12   39  0.697   
18  2015-06   80  43~188                good     43    57   11   35  0.697   
19  2015-07   84  31~171                good     39    57   11   31  0.658   

     O3  
0    49  
1    58  
2    73  
3   100  
4   108  
5   142  
6   118  
7   112  
8   112  
9   117  
10  119  
11   84  
12   47  
13   56  
14   81  
15  102  
16  122  
17  126  
18  129  
19  134

In [52]:
air_qa.describe()

AQI       PM2.5        PM10         SO2         NO2          CO  \
count  132.000000  132.000000  132.000000  132.000000  132.000000  132.000000   
mean    73.871212   36.727273   52.833333   10.181818   38.030303    0.715144   
std     15.891524   16.887635   20.540099    6.356842   12.959679    0.152789   
min     45.000000    3.000000    6.000000    4.000000   11.000000    0.442000   
25%     64.000000   24.750000   36.750000    6.000000   29.000000    0.605250   
50%     72.000000   33.000000   51.000000    8.000000   36.000000    0.697000   
75%     81.000000   45.000000   63.250000   12.000000   47.250000    0.781000   
max    158.000000  125.000000  140.000000   48.000000   71.000000    1.373000   

               O3  
count  132.000000  
mean    99.825758  
std     26.670593  
min     34.000000  
25%     81.000000  
50%    101.500000  
75%    121.000000  
max    177.000000

## no null values in air quality

In [53]:
air_qa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          132 non-null    object 
 1   AQI            132 non-null    int64  
 2   scope          132 non-null    object 
 3   Quality Grade  132 non-null    object 
 4   PM2.5          132 non-null    int64  
 5   PM10           132 non-null    int64  
 6   SO2            132 non-null    int64  
 7   NO2            132 non-null    int64  
 8   CO             132 non-null    float64
 9   O3             132 non-null    int64  
dtypes: float64(1), int64(6), object(3)
memory usage: 10.4+ KB


## need to combine air quality data with the dataframe for other weather conditions temperature,pressure etc

# converting air quality date column 'month' to datetime

In [57]:
air_qa['month']=pd.to_datetime(air_qa['month'],format='%Y-%m')

In [61]:
monthly_data.reset_index(inplace=True)

In [62]:
monthly_data.head(50)

Timestamp  Wind Direction  Wind Speed  Temperature   Humidity     Pressure
0  2014-01-31       56.271436   13.200298     7.409946  63.084547  1026.000000
1  2014-02-28       32.504738   16.504820     6.581845  73.610214  1026.000000
2  2014-03-31      104.239511   15.425806    12.075269  62.895161  1020.000000
3  2014-04-30      103.216295   14.495000    16.290278  68.023611  1015.000000
4  2014-05-31      143.646904   15.435694    22.100806  60.005312  1011.000000
5  2014-06-30      111.282868   13.017350    23.987500  72.573014  1006.000000
6  2014-07-31      141.871173   14.988470    28.262168  76.621637  1005.000000
7  2014-08-31       86.604688   12.489610    26.812500  81.154265  1006.000000
8  2014-09-30       67.442191   14.910255    24.685417  77.301536  1012.000000
9  2014-10-31       46.934248   14.709033    20.584677  67.791978  1020.000000
10 2014-11-30       22.978962   13.523585    15.048611  69.760548  1022.000000
11 2014-12-31      321.707653   13.392285     6.173387  56.492106  1027.000000
12 2015-01-31      350.816351   14.017742     6.755376  67.060277  1026.000000
13 2015-02-28       31.505447   13.483929     7.470238  69.452081  1026.000000
14 2015-03-31       97.804063   13.117742    11.559140  70.188172  1020.000000
15 2015-04-30       98.688934   14.844970    16.516667  66.089136  1015.000000
16 2015-05-31      120.180684   14.755013    21.140457  70.184755  1011.000000
17 2015-06-30      133.066106   12.650000    24.494444  80.559722  1006.000000
18 2015-07-31      121.365065   14.012914    27.334677  76.939729  1005.000000
19 2015-08-31      106.734926   15.222893    28.278226  74.562918  1006.484719
20 2015-09-30       33.803872   13.737320    24.758333  72.401389  1012.535972
21 2015-10-31       59.835422   11.629712    20.051747  68.997221  1019.162908
22 2015-11-30        2.074025   12.771846    14.262500  80.347222  1022.647222
23 2015-12-31      336.663785   12.517079     8.315860  71.505376  1027.084677
24 2016-01-31        0.633655   14.380645     5.092742  70.451613  1028.312941
25 2016-02-29       21.884545   14.768212     7.720546  60.569575  1026.217511
26 2016-03-31       93.324441   14.486253    11.875000  65.938172  1021.075955
27 2016-04-30      117.928794   14.186574    17.415278  72.887500  1012.801466
28 2016-05-31      109.868809   13.937141    21.116935  75.223118  1010.762097
29 2016-06-30      160.081963   12.965000    24.608333  83.154167  1006.659722
30 2016-07-31      149.513891   13.596142    30.471774  74.991935  1005.090028
31 2016-08-31       98.914443   15.236746    30.200269  68.591398  1004.936483
32 2016-09-30       61.549127   13.508446    25.425000  74.836111  1011.166667
33 2016-10-31       36.550550   15.309677    20.998656  79.197581  1017.181452
34 2016-11-30       18.670330   13.008616    13.888889  77.580556  1022.501389
35 2016-12-31       17.860253   13.354839     9.775538  72.872312  1025.673387
36 2017-01-31       25.959305   14.168991     7.353495  71.409946  1026.361358
37 2017-02-28       47.242820   14.105357     7.531250  63.727679  1025.486254
38 2017-03-31       57.243769   14.254839    10.913978  66.112903  1020.637589
39 2017-04-30      135.809947   14.525000    17.997222  64.750000  1013.939358
40 2017-05-31      136.318859   15.412378    22.612903  65.555108  1011.422043
41 2017-06-30      128.793304   12.580609    24.472222  76.756944  1007.376389
42 2017-07-31      169.912515   14.433871    32.541667  65.850806  1005.797631
43 2017-08-31      156.190215   14.758828    30.231183  73.317204  1004.344086
44 2017-09-30       67.106374   14.626622    25.150000  76.097222  1011.365278
45 2017-10-31       19.819450   16.556391    19.424731  73.219086  1019.682796
46 2017-11-30       28.157774   13.208053    14.101389  69.033333  1023.038889
47 2017-12-31       14.651735   12.509472     7.624328  61.160761  1027.321891
48 2018-01-31        4.150241   14.496774     4.770161  72.461022  1025.341954
49 2018-02-28       26.185395   14.078571     6.007440

## there are days in date column of both tables so removing them to join them on month and year

In [63]:
monthly_data['Year'] = monthly_data['Timestamp'].dt.year
monthly_data['Month'] = monthly_data['Timestamp'].dt.month
air_qa['Year'] = air_qa['month'].dt.year
air_qa['Month'] = air_qa['month'].dt.month



## finally merging the data based on month and year to get complete dataset 

In [64]:
combined_df = pd.merge(monthly_data, air_qa, on=['Year', 'Month'], how='inner')

In [65]:
combined_df = combined_df.drop(columns=['Timestamp', 'month'])

## Final data

In [67]:
combined_df.tail(20)

Wind Direction  Wind Speed  Temperature   Humidity     Pressure  Year  \
111      120.373421   15.665556    17.900000  64.061111  1014.481944  2023   
112      133.907690   14.713710    22.353495  68.512097  1011.142473  2023   
113      154.543186   12.368889    26.165278  75.966667  1006.465139  2023   
114      167.708623   14.104301    30.043011  78.645161  1005.523253  2023   
115       66.570301   13.127151    29.455645  74.611559  1004.669355  2023   
116       61.023578   11.443333    26.390278  78.404167  1011.838750  2023   
117       48.550879   11.081720    21.066801  67.419355  1019.397715  2023   
118      321.215866   13.935556    14.887500  60.848611  1023.186250  2023   
119      329.751841   12.502151     7.537634  65.294355  1026.823253  2023   
120      356.159741   12.846774     6.594086  66.689516  1026.935215  2024   
121      347.379127   15.516667     6.846264  75.396552  1024.964511  2024   
122       96.409317   13.397849    12.271909  67.778226  1018.720565  2024   
123       96.652195   13.155000    17.545833  73.590278  1012.629028  2024   
124      124.164003   14.695968    21.677419  64.368280  1013.188710  2024   
125      164.679932   12.019444    24.929167  77.311111  1007.366111  2024   
126      178.374792   17.010753    31.580645  69.548387  1004.698522  2024   
127      135.622965   12.851344    31.934140  67.279570  1005.059812  2024   
128       84.487889   15.679167    28.128889  76.018056  1009.637222  2024   
129       18.242554   14.990054    20.412634  74.865591  1019.106183  2024   
130       14.797232   14.293194    15.417639  67.195833  1023.228750  2024   

     Month  AQI   scope Quality Grade  PM2.5  PM10  SO2  NO2     CO   O3  
111      4   81  45~188          good     33    83    7   29  0.663  114  
112      5   78  38~165          good     28    50    7   27  0.584  129  
113      6   79  28~146          good     26    34    7   25  0.623  131  
114      7   64  18~136          good     18    29    6   20  0.558  107  
115      8   62  24~163          good     17    30    7   18  0.603  107  
116      9   58  25~112          good     19    29    7   25  0.670  101  
117     10   64  41~116          good     24    41    7   34  0.655  115  
118     11   64  34~115          good     33    61    8   42  0.743   83  
119     12   77  30~202          good     49    75    8   53  0.884   34  
120      1   85  33~165          good     57    81    8   53  0.926   63  
121      2   63  27~214          good     40    53    7   26  0.766   80  
122      3   69  44~102          good     37    64    7   41  0.700   97  
123      4   59  38~110          good     27    43    7   31  0.630  103  
124      5   69  41~106          good     21    36    7   23  0.561  121  
125      6   65  32~142          good     20    24    7   24  0.597  111  
126      7   71  26~119          good     19    31    7   20  0.542  120  
127      8   79  36~165          good     19    30    7   18  0.616  133  
128      9   46  20~114     excellent     14    25    7   18  0.570   83  
129     10   53  32~100          good     20    31    7   27  0.645   95  
130     11   45   45~45     excellent      3     6    6   11  0.500   90

## converting to excel file

In [69]:
file_path = f'C:/Users/hp/Downloads/timeserie/combined_data.xlsx'  # You can provide a full path if needed

# Save the DataFrame to an Excel file
combined_df.to_excel(file_path, index=False)

print(f"DataFrame saved to {file_path}")

DataFrame saved to C:/Users/hp/Downloads/timeserie/combined_data.xlsx
